File: C++/collatz.cpp
// Hannah Gulle
// Project 3 -- The Collatz Conjecture
// Given a starting integer within the compiler max integer range,
// the Collatz sequence for the beginning value will be computed 
// and the sequence size printed to the screen.
// Additionally, for the max integer range given gcc 6.4.0
// restrictions in c++, the 10 numbers producing the longest 
// Collatz Sequences will be determined.
// Consider c++MaxIntegerValues.txt for gcc 6.4.0 integer
// range restrictions, as considered in this program.

#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

// Collatz Struct holds sequence start and sequence length
struct Collatz{
	unsigned long long int start;
	int length;
};

// Sorting Methods
bool startSort(Collatz const& lhs, Collatz const& rhs);
bool lengthSort(Collatz const& lhs, Collatz const& rhs);

// Collatz Method
vector<unsigned long long int> collatz( unsigned long long int start );

// Main Program
int main(){

	// Retrieves the highest starting value from the user
	unsigned long long int end;
	cout << "Enter Highest Starting Value: ";
	cin >> end;

	// collatz struct vector holds all objects within computation starting range
	vector<Collatz> all;
	// vector holds all sequence values for a given start
	vector<unsigned long long int> sequence;

	// determines and constructs the collatz structs to fill the collatz vector
	// for each starting value and sequence length after collatz construction
	int size;
	for(unsigned long long int i = 2; i < end; i++){
		sequence = collatz(i);
		size = sequence.size()-1;
		
		Collatz c = {i, size};
		all.push_back(c);
		
		sequence.clear();	

	}

	cout << "Top 10 Starting Values After Sorting by Length" << endl;

	// Sorts collatz structs by length of sequence
	sort( all.begin(), all.end(), &lengthSort );

	// Outputs the top 10 starting values after Sorting by length
	vector<Collatz> topTen;
	int top = 1;
	for(int i = all.size()-1; i > 1; i--){

		if( all[i].length > all[i-1].length ){
			topTen.push_back(all[i]);
			cout << all[i].start << " " << all[i].length << endl;
			top++;
		}
		if( top > 10 ){
			break;
		}
	}

	// Sorts collatz struct vector of top 10 objects by starting value
	sort( topTen.begin(), topTen.end(), &startSort );

	cout << "\n Top 10 Starting Values After Sorting By Start" << endl;

	// outputs top 10 starting values after sorting by start
	for( int i = 9; i > 0; i-- ){
		cout << topTen[i].start << " " << topTen[i].length << endl;
	}

	return 0;
}

// sorting function: sorts by collatz object start value
bool startSort(Collatz const& lhs, Collatz const& rhs ){
	return lhs.start < rhs.start;
}

// sorting function: sorts by collatz object length value
bool lengthSort(Collatz const& lhs, Collatz const& rhs){
	return lhs.length < rhs.length;
}

// calculates the collatz sequence for a given start value
vector<unsigned long long int> collatz( unsigned long long int start ){

	vector<unsigned long long int> sequence;
	sequence.push_back(start);

	while( start != 1 ){

		if( start % 2 == 0 ){
			start /= 2;
		}
		else{
			start = ( 3 * start ) + 1;
		}
		sequence.push_back(start);
	}

	return sequence;
}

File: F2015/collatz.f90
! Hannah Gulle
! Computes the top 10 collatz starting integers within the given user range

program runCollatz

! c like struct holds the sequence start and sequence length
type collatz
        integer::start
        integer::length
end type

! collatz types for adding objects to the array of all collatz
! struct like objects
type(collatz) :: c, temp

! Dynamically allocated array that holds all collatz type structs
type(collatz), dimension(:), allocatable :: allseq

! Collatz Sequence Finding Variables
integer :: endval       ! Highest Starting Value
integer :: seqSize      ! Length of Sequence
integer :: start, top   ! Starting Value
integer :: i,j          ! Iterative Values

integer :: mem_stat     ! Allocation Error Code

! Holds the top ten collatz struct types
type(collatz), dimension(10) :: topTen 

! Retrieves the highest starting integer from the user
write(*,*) "Please Enter the Highest Starting Integer "
read(*,*) endval

! Dynamically allocates the collatz type array for the 
! given range of values
allocate( allseq( endval ), stat=mem_stat)
if(mem_stat /= 0) STOP "Memory Allocation Error"

! Determines and Constructs the array of collatz type structs
! for the user given range
do 10 i=2, endval, 1

        start = i
        seqSize = 0
        
        do while (start /= 1 )
                if( mod(start,2) == 0 ) then
                        start = start / 2
                else
                        start = (3*start)+1
                endif
                seqSize = seqSize + 1
        enddo

        c = collatz( i, seqSize )
        allseq((i-1)) = c
10 continue

! Sort all collatz objects by collatz length
do i=2, size(allseq)
        j = i - 1
        temp = allseq(i)
        do while (j >= 1 .and. allseq(j)%length > temp%length)
                allseq(j+1) = allseq(j)
                j = j - 1
        end do
        allseq(j+1) = temp
end do

print *, "Top 10 Starting Values After Sorting by Length"
! Output top 10 collatz start values after sorting by length
top = 1

do i=size(allseq), 1, -1
        if( allseq(i)%length > allseq(i-1)%length ) then
                topTen(top) = allseq(i)
                print *, allseq(i)%start, allseq(i)%length
                top = top + 1
        end if
        if( top > 10 ) then
                EXIT
        end if

end do

! Sort top 10 collatz objects by starting value
do i=2, 10
        j = i - 1
        temp = topTen(i)
        do while (j >= 1 .and. topTen(j)%start > temp%start)
                topTen(j+1) = topTen(j)
                j = j - 1
        end do
        topTen(j+1) = temp
end do

print *, "Top 10 Starting Values After Sorting by Start"
! Output top 10 after sorting by start
do i=10, 1, -1
        print *, topTen(i)%start, topTen(i)%length
end do
end program runCollatz
